cdc/scheduler/internal/tp/scheduler_drain_capture.go

```diff
@@ -0,0 +1,170 @@
+		return false
+	}
+
+	d.target = target
+	return true
+}
+
+func (d *drainCaptureScheduler) Schedule(
+	_ model.Ts,
+	_ []model.TableID,
+	captures map[model.CaptureID]*model.CaptureInfo,
+	replications map[model.TableID]*ReplicationSet,
+) []*scheduleTask {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if d.target == captureIDNotDraining {
+		return nil
+	}
+
+	var availableCaptureCount int
+	for id := range captures {
+		if id == d.target {
+			availableCaptureCount++
+		}
+	}
+
+	// this may happen when inject the target, there is at least 2 alive captures
+	// but when schedule the task, only owner alive.
+	if availableCaptureCount == 0 {
+		log.Warn("tpscheduler: drain capture scheduler ignore drain target capture, "+
+			"since cannot found destination captures",
+			zap.String("target", d.target), zap.Any("captures", captures))
+		d.target = captureIDNotDraining
+		return nil
+	}
+
+	// victims record all table instance should be dropped from the target capture
+	victims := make([]model.TableID, 0)
+	captureWorkload := make(map[model.CaptureID]int)
+	for tableID, rep := range replications {
+		if rep.State != ReplicationSetStateReplicating {
+			// only drain the target capture if all tables is replicating,
+			log.Debug("tpscheduler: drain capture scheduler skip this tick,"+
+				"not all table is replicating",
+				zap.String("target", d.target),
+				zap.Any("replication", rep))
+			return nil
+		}
+
+		if rep.Primary == d.target {
+			victims = append(victims, tableID)
+		}
+
+		// only calculate workload of other captures not the drain target.
+		if rep.Primary != d.target {
+			captureWorkload[rep.Primary] += 1
+		}
+	}
+
+	if len(victims) == 0 {
+		log.Info("tpscheduler: drain capture scheduler finished, since no table",
+			zap.String("target", d.target), zap.Any("captures", captures))
+		d.target = captureIDNotDraining
+		return nil
+	}
+
+	for captureID, w := range captureWorkload {
+		captureWorkload[captureID] = randomizeWorkload(d.random, w)
+	}
+
+	// for each victim table, find the target for it
+	moveTables := make([]moveTable, 0, len(victims))
+	for _, tableID := range victims {
+		target := ""
+		minWorkload := math.MaxInt64
+
+		for captureID, workload := range captureWorkload {
+			if workload < minWorkload {
+				minWorkload = workload
+				target = captureID
+			}
+		}
+
+		if minWorkload == math.MaxInt64 {
+			log.Panic("tpscheduler: drain capture meet unexpected min workload " +
+				"when try to the the target capture")
+		}
+
+		moveTables = append(moveTables, moveTable{
+			TableID:     tableID,
+			DestCapture: target,
+		})
+
+		captureWorkload[target] = randomizeWorkload(d.random, minWorkload+1)
+	}
+
+	accept := func() {
+		d.mu.Lock()
+		defer d.mu.Unlock()
+		log.Info("tpscheduler: drain capture task accepted",
+			zap.String("target", d.target))
+	}
+
+	// todo: revise this if we have to control the concurrency.
+	task := &scheduleTask{
+		burstBalance: &burstBalance{MoveTables: moveTables},
```

Comment:
I think it use be a normal move tables task. We need to limit total number of concurrent scheduling tasks.